Add new int8 scalar quantization to HNSW codec

benwtrent opened the pull request:
As with most codec changes, this is an eye popping number of LoC and the design isn't finished yet. 

Initial benchmarking (utilizing non-normalized cohere embeddings + max-inner product, which is a particularly difficult case for naive quantization), I get 10-20% faster search and ~4x smaller storage used for the search (I am keeping the raw vectors around...we can debate if we want to do that).

Recall@10 with 100 fanout = 0.804
Recall@100 with 200 fanout = 0.9.

I am reaching the point where the design needs to be finalized and I wanted to reachout for feedback.

Some design discussion points that I am unsure about are:

 - Do we want to have a new "flat" vector codec that HNSW (or other complicated vector indexing methods), can use? Detractor here is that now HNSW codec relies on another pluggable thing that is a "flat" vector index (just provides mechanisms for reading, writing, merging vectors in a flat index).
 - Should "quantization" just be a thing that is provided to vector codecs? The main detractor here is future scalar quantization could easily be added (like int4 or even binary). 
 - Should the "quantizer" keep the raw vectors around itself? Or rely on some external party to provide them (in this case, I an relying on the HNSW codec)?

benwtrent committed:
AMEND ME INITIAL QUANTIZATION CODE

benwtrent committed:
SQUASH AND REMOVE

benwtrent committed:
CLEAN

jimczi committed:
Streamlined vector retrieval and scoring in HNSW Builder/Searcher

This PR involves the refactoring of the HNSW builder and searcher, aiming to create an abstraction for the random access and vector comparisons conducted during graph traversal.

The newly added RandomVectorScorer provides a means to directly compare ordinals, eliminating the need to expose the raw vector primitive type.
This scorer takes charge of vector retrieval and comparison during the graph's construction and search processes.

The primary purpose of this abstraction is to enable the implementation of various strategies.
For example, it opens the door to constructing the graph using the original float vectors while performing searches using their quantized int8 vector counterparts.

jimczi committed:
address review

jimczi committed:
add javadoc

jimczi committed:
require non null scorer provider

jimczi committed:
restore import

jimczi committed:
tidy

jimczi committed:
deep copy of the provided query array

jimczi committed:
reuse query instead of copying the value

jimczi committed:
rename param to avoid confusion

jimczi committed:
tidy

benwtrent committed:
iter

jimczi committed:
address review comments

jimczi committed:
remove unused variable

jimczi committed:
tidy

jimczi committed:
Merge branch 'main' into graph_similarity_scorer

benwtrent committed:
iter

benwtrent committed:
reverting all unnecessary changes

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

jimczi committed:
apply review comments

jimczi committed:
fix format

benwtrent committed:
squash

benwtrent committed:
iter

benwtrent committed:
iter

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
Merge remote-tracking branch 'jimczi/graph_similarity_scorer' into feature/add-scalar-quantization

benwtrent committed:
adding new lucene98 codec

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
fixing merge

benwtrent committed:
fixing codec bwc and such

benwtrent committed:
More test fixes

benwtrent committed:
fixing tests and backwards compat

benwtrent committed:
fixing tests and backwards compat

benwtrent committed:
rm lucene95 in favor of backward_codec

benwtrent committed:
improve quantization

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
post merge fixup

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,851 @@
+    return mergeAndRecalculateQuantiles(mergeState, fieldInfo, quantile);
+  }
+
+  @Override
+  public IndexInput mergeOneField(
+      FieldInfo fieldInfo, MergeState mergeState, ScalarQuantizationState mergedQuantizationState)
+      throws IOException {
```
This type of interaction tells me that my current design is seriously flawed. 

During segment merging, we need access to the merged quantized values. That merging may require requantization and may require us calculating new quantiles, etc.

Then, the caller needs access to the newly merged quantized vectors. However, since these are done within a writer context, I could not easily see a way to return a "reader" interface that was also closable (we don't want to keep `quantizedVectorDataInput` open).

@jpountz @jimczi what do you think here?

This indicates to me that merging needs to be done differently. Or quantization isn't a "codec" type thing, but just a set of functions that provide the tools to read/write/merge the correct things (and consequently, wouldn't have logic like this).

jimczi replied:
Quantisation as a codec makes sense to me but not as an additive layer on top of the HNSW one. If we want to follow this path I think we need to separate the concern and make another codec responsible for writing the flat index in HNSW. That's your suggestion below and I think it's much cleaner.
The question is whether we want to generalise the feature on the first attempt. We could also start by adding a quantization option inlined with the HNSW codec.

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,851 @@
+  }
+
+  @Override
+  public void flush(int maxDoc, Sorter.DocMap sortMap) throws IOException {
```
Partially duplicating the KnnVectorWriter codec interface tells me that this API is just different enough that we need to rethink it.

Basically, I am not 100% sure if we want to add a new "FlatKnnCodec" that quantization satisfies. Thus allowing HNSW to wrap another codec (this gets hairy as, what if somebody wants to wrap HNSW with HNSW?!?!?!).

OR, we have SOME other way of exposing these functions. 

NOTE: We will want to add new quantization methods in the future. And future quantization methods will probably want to read from other quantized kinds (e.g. int8 being read into an int4 quantizer and then writing out int4 quantized values).

jimczi replied:
I am in favour of an HNSW codec that wraps another codec to store/access/score the underlying vectors. That's the most flexible way to introduce quantisations. I also like the fact that a "flat" codec used without HNSW can be used to rescore vectors or as a way to keep the original vectors in a different field.

jimczi replied:
The main downside of this approach is that we expose the float and byte vector values as doc values in the reader/writer. That prevents any transformation in the codec because the API forces us to expose the original values in a sense. We need to decide whether exposing these values on the reader side is the expected behaviour. The API was not meant for lossy compression.

jpountz reviewed (commented):
This is exciting stuff! I have questions (beware that I know little about vector search and only skimmed over the PR).
 - Do we know why search is faster? Is it mostly because working on the quantized vectors requires a lower memory bandwith?
 - Do you know how recall degrades compared to without quantization? I saw the numbers you shared but I don't have a good sense of what recall we usually had until now.
 - I don't feel great about the logic that merges quantiles at merge time and only requantizes if the merged quantiles don't differ too much from the input quantiles. It feels like quantiles could slowly change over multiple merging rounds and we'd end up in a state where the quantized vectors would be different from requantizing the raw vectors with the quantization state that is stored in the segment, which feels wrong. Am I missing something?
 - Related to the above, it looks like we ignore deletions when merging quantiles. It would probably be ok in practice most of the time but I worry that there might be corner cases?

> Do we want to have a new "flat" vector codec that HNSW (or other complicated vector indexing methods), can use? Detractor here is that now HNSW codec relies on another pluggable thing that is a "flat" vector index (just provides mechanisms for reading, writing, merging vectors in a flat index).

I don't have a strong opinion on this. Making it a codec though has the downside that it would require more files since two codecs can't write to the same file. Maybe having utility methods around reading/writing flat vectors is good enough?

> Should "quantization" just be a thing that is provided to vector codecs? 

I might be misunderstanding the question, but to me this is what the byte[] encoding is about. And this quantization that's getting added here is more powerful because it's adaptative and will change over time depending on what vectors get indexed or deleted? If it needs to adapt to the data then it belongs to the codec. We could have utility code to make it easier to write codecs that quantize their data though (maybe this is what your question suggested?).

> Should the "quantizer" keep the raw vectors around itself?

My understanding is that we have to, as the accuracy of the quantization could otherwise degrade over time  in an unbounded fashion.

benwtrent reviewed (commented):
@jpountz @jimczi I highlighted two glaring issues in this giant sea of a change-log.

benwtrent commented:
> Do we know why search is faster? Is it mostly because working on the quantized vectors requires a lower memory bandwi[d]th?

Search is faster in two regards:

 - PanamaVector allows for more `byte` actions to occur at once than `float32` (should be major)
 - Reading `byte[]` off of a buffer doesn't require decoding floats (very minor change)

IMO, we should be seeing WAY better search numbers. I need to do more testing to triple check.

> Do you know how recall degrades compared to without quantization? I saw the numbers you shared but I don't have a good sense of what recall we usually had until now.

++ I want to graph the two together to compare so its clearer.



> I don't feel great about the logic that merges quantiles at merge time and only requantizes if the merged quantiles don't differ too much from the input quantiles. It feels like quantiles could slowly change over multiple merging rounds and we'd end up in a state where the quantized vectors would be different from requantizing the raw vectors with the quantization state that is stored in the segment, which feels wrong. Am I missing something?

The quantization buckets could change slightly overtime, but since we are bucketing `float32` into `int8`, the error bounds are comparatively large. 

The cost of requantization is almost never worth it. In my testing, quantiles over random data from the same data set shows that segments differ by only around `1e-4`, which is tiny and shouldn't require requantization.

@tveasey helped me do some empirical analysis here and can provide some numbers.


> Related to the above, it looks like we ignore deletions when merging quantiles. It would probably be ok in practice most of the time but I worry that there might be corner cases?

A corner case in what way? That we potentially include deletions when computing quantiles or if re-quantization is required?

We can easily exclude them as conceptually, the "new" doc (if it were an update) would exist in another segment. It could be we are double counting a vector and we probably shouldn't do that.

> > Do we want to have a new "flat" vector codec that HNSW (or other complicated vector indexing methods), can use? Detractor here is that now HNSW codec relies on another pluggable thing that is a "flat" vector index (just provides mechanisms for reading, writing, merging vectors in a flat index).

> I don't have a strong opinion on this. Making it a codec though has the downside that it would require more files since two codecs can't write to the same file. Maybe having utility methods around reading/writing flat vectors is good enough?

Utility methods are honestly what I am leaning towards. Its then a discussion around how a codec (like HNSW) is configured to use it.

> > Should "quantization" just be a thing that is provided to vector codecs?

> I might be misunderstanding the question, but to me this is what the byte[] encoding is about. And this quantization that's getting added here is more powerful because it's adaptative and will change over time depending on what vectors get indexed or deleted? If it needs to adapt to the data then it belongs to the codec. We could have utility code to make it easier to write codecs that quantize their data though (maybe this is what your question suggested?).

Yeah, it needs to adapt over time. There are adverse cases (indexing vectors sorted by relative clusters is one) that need to be handled. But, they can be handled easily at merge time by recomputing quantiles and potentially re-quantizing.

> > Should the "quantizer" keep the raw vectors around itself?

> My understanding is that we have to, as the accuracy of the quantization could otherwise degrade over time in an unbounded fashion.

After a period of time, if vectors are part of the same corpus and created via the same model, the quantiles actually level out and re-quantizing will rarely or never occur since the calculated quantiles are statistically equivalent. Especially given the binning into `int8`.

uschindler commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,209 @@
+  public float calculateVectorOffset(byte[] vector, VectorSimilarityFunction similarityFunction) {
+    if (similarityFunction != VectorSimilarityFunction.EUCLIDEAN) {
+      int sum = 0;
+      for (byte b : vector) {
```
Can't we use VectorUtil here for SIMD dotProduct?

uschindler replied:
Ah sorry it just sums up. But we could add this to VectorUtil...

rmuir replied:
summing bytes across array like this should work with autovectorization, or its seriously broke. there is no pesky floating point order of operations restriction.

benwtrent replied:
@rmuir, exactly. Since it isn't floating point addition, I didn't think it necessary for VectorUtil to get involved.

benwtrent committed:
fixing corrective constant code

benwtrent committed:
merging scalar quantization into hnsw

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsWriter.java:
```diff
@@ -0,0 +1,1170 @@
+      if (quantizedVectorsWriter != null) {
+        quantizationState =
+            quantizedVectorsWriter.mergeOneField(segmentWriteState, fieldInfo, mergeState);
+        if (quantizationState != null) {
+          quantizedVectorDataOffsetAndLength = new long[2];
+          quantizedVectorDataOffsetAndLength[0] = quantizedVectorData.alignFilePointer(Float.BYTES);
+          quantizationDataInput =
+              segmentWriteState.directory.openInput(
+                  quantizationState.tempVectorFileName, segmentWriteState.context);
+          quantizedVectorData.copyBytes(
+              quantizationDataInput, quantizationDataInput.length() - CodecUtil.footerLength());
+          quantizedVectorDataOffsetAndLength[1] =
+              quantizedVectorData.getFilePointer() - quantizedVectorDataOffsetAndLength[0];
+          CodecUtil.retrieveChecksum(quantizationDataInput);
+        }
```
For merging and reading, I need to:

 - merge the quantile information and determine if we should recalculate quantiles and requantize
 - Merge the quantized files into a single file
 - provide some way to then read the merged file using the merged quantization settings for graph building.

This seemed like the best way at the time (return the temp-file name and allow the hnsw writer read and delete it later).

benwtrent reviewed (commented):
@jpountz @jimczi updated the API a bit, what do y'all think?

jimczi commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsWriter.java:
```diff
@@ -0,0 +1,1170 @@
+          CodecUtil.retrieveChecksum(quantizationDataInput);
+        }
+      }
+      // write the vector data to a temporary file
```
We don't need to write the raw vectors in a temporary file if we build the graph with the quantised vectors? This part could be extracted in a separate function that writes the raw and quantised vectors and returns a "closeable" `RandomVectorScorerSupplier`.

benwtrent replied:
FYI, I think we will always need a temporary file for SOMETHING. Because we need a way to read the vectors we just wrote and opening & closing the `vec` or `veq` file on every field merge seems like too much.

jimczi replied:
it looks good, thanks for adding the comments

jimczi commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/package-info.java:
```diff
@@ -0,0 +1,437 @@
+ *   <li>In version 9.5, HNSW graph connections were changed to be delta-encoded with vints.
+ *       Additionally, metadata file size improvements were made by delta-encoding nodes by graph
+ *       layer and not writing the node ids for the zeroth layer.
+ *   <li>In version 9.8, Vector scalar quantization support was added. Allowing the HNSW vector
```
```suggestion
 *   <li>In version 9.9, Vector scalar quantization support was added. Allowing the HNSW vector
```

jimczi reviewed (commented):
I like the simpler way of putting quantization in just one codec. I've noticed a couple of small tweaks we can make when using quantization instead of raw vectors, but I think it's good to take it out of draft mode now! 😄

benwtrent commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsWriter.java:
```diff
@@ -0,0 +1,1170 @@
+    int count = docsWithField.cardinality();
+    meta.writeInt(count);
+    if (isQuantized) {
+      writeOrdToDoc(quantizedVectorData, docsWithField, count, maxDoc);
```
@jimczi wouldn't this require opening the raw vectors file? Maybe I misunderstand how ordToDoc translations occur. Is it completely read onto the heap or do we keep that file open?

jimczi replied:
You mean when reading? The raw vectors file is already opened so not sure I follow. 
The `DirectMonotonicReader` is off-heap but my point is that we can reuse the same mapping since ord to doc is not dependent of the quantisation. It seems weird that we need to write/read this info in two different places.

benwtrent replied:
> The raw vectors file is already opened so not sure I follow.

I mean, not have to seek in the file. I would have thought seeking the file would page it into cache.

jimczi replied:
Well if we're using the same file ;). From what I understand we write the same data in two different places so the cache won't be shared.

benwtrent replied:
> From what I understand we write the same data in two different places so the cache won't be shared.

Yes, I write the offsets at the end of `veq` so that if that file is paged into memory, we can access those offsets. 

I would expect searching sparse vectors in `veq` and then looking for offsets in `vec` would cause more thrashing no?

benwtrent replied:
IMO, this is a stop-gap, it would be really good to separate concerns around things that store the vectors vs. things that index them :/

benwtrent committed:
addressing pr comments, clean up

benwtrent committed:
Merge branch 'main' into feature/add-scalar-quantization

benwtrent committed:
updating package info

benwtrent commented:
I was doing some performance testing and was getting weird results. Quantization search and indexing build were marginally better or exactly the same. 

Attached a zip of async-profiler outputs for `wall` & `cpu`, the `-95-` files are quantized to the `.95` quantile and are using `dotProduct(byte[], byte[])`

[Archive 2.zip](https://github.com/apache/lucene/files/12797059/Archive.2.zip)

Digging into why performance wasn't any better, I used Robert's vector bench library to see how my macbook did with our `byte` implementation vs `float` and WOW `byte[]` seems REALLY slow. 

```
FloatDotProductBenchmark.dotProductNew     768  thrpt    5  21.781 ± 0.254  ops/us
FloatDotProductBenchmark.dotProductNew    1024  thrpt    5  15.091 ± 0.217  ops/us
FloatDotProductBenchmark.dotProductNew    2048  thrpt    5   8.233 ± 0.023  ops/us
BinaryDotProductBenchmark.dotProductNew     768  thrpt    5  8.041 ± 0.108  ops/us
BinaryDotProductBenchmark.dotProductNew    1024  thrpt    5  6.085 ± 0.133  ops/us
BinaryDotProductBenchmark.dotProductNew    2048  thrpt    5  3.054 ± 0.123  ops/us
```

768 dims is not even as fast as 2048 floats! This is counter intuitive to me, SIMD should be able to process MORE information per cycle, even on a MacBook with only 128-bitsize.

I am going to have to move my testing off of my machine to something with a different CPU. But in parallel, we should figure out why `byte[]` is so bad when compared to `float[]` (my guess is just because we just don't have the correct intrinsics for now...)

tveasey commented:
I saw go by since I’m mentioned on the PR. It seems like Java can’t lay out byte vectors properly: https://stackoverflow.com/questions/14531235/in-java-is-it-more-efficient-to-use-byte-or-short-instead-of-int-and-float-inst and so incurs type coercion penalties. But I guess Panama does this properly, so you should have a workaround.

> On 3 Oct 2023, at 22:28, Benjamin Trent ***@***.***> wrote:
> 
> 
> I was doing some performance testing and was getting weird results. Quantization search and indexing build were marginally better or exactly the same.
> 
> Attached a zip of async-profiler outputs for wall & cpu, the -95- files are quantized to the .95 quantile and are using dotProduct(byte[], byte[])
> 
> Archive 2.zip <https://github.com/apache/lucene/files/12797059/Archive.2.zip>
> Digging into why performance wasn't any better, I used Robert's vector bench library to see how my macbook did with our byte implementation vs float and WOW byte[] seems REALLY slow.
> 
> FloatDotProductBenchmark.dotProductNew     768  thrpt    5  21.781 ± 0.254  ops/us
> FloatDotProductBenchmark.dotProductNew    1024  thrpt    5  15.091 ± 0.217  ops/us
> FloatDotProductBenchmark.dotProductNew    2048  thrpt    5   8.233 ± 0.023  ops/us
> BinaryDotProductBenchmark.dotProductNew     768  thrpt    5  8.041 ± 0.108  ops/us
> BinaryDotProductBenchmark.dotProductNew    1024  thrpt    5  6.085 ± 0.133  ops/us
> BinaryDotProductBenchmark.dotProductNew    2048  thrpt    5  3.054 ± 0.123  ops/us
> 768 dims is not even as fast as 2048 floats! This is counter intuitive to me, SIMD should be able to process MORE information per cycle, even on a MacBook with only 128-bitsize.
> 
> I am going to have to move my testing off of my machine to something with a different CPU. But in parallel, we should figure out why byte[] is so bad when compared to float[] (my guess is just because we just don't have the correct intrinsics for now...)
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/apache/lucene/pull/12582#issuecomment-1745753817>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABZ5MP65P3QS7ILCV3KHHUDX5R7PRAVCNFSM6AAAAAA5CAEGCSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONBVG42TGOBRG4>.
> You are receiving this because you were mentioned.
>

benwtrent committed:
fixing offset calculation and scalar requantization

benwtrent committed:
fixing writer

benwtrent committed:
removing unnecessary change

benwtrent commented:
I did some benchmarks with the current change, using 400k Cohere embeddings searching over 1k vectors. This was on GCP `c3-standard-8` (Intel Sapphire Rapids), on this machine `byte[]` comparisons run in similar time to `float[]`.

Index buffer: 128MB
Force merged to one segment.

I noticed that force-merging was about 35% faster.

For query latency:

| query         | recall | latency |
|---------------|--------|---------|
| raw           | 0.845  | 0.31ms  |
| quantized @10 | 0.800  | 0.23ms  |
| quantized @20 | 0.862  | 0.29    |

Disk usage:
 - raw: 1172MB
 - quantized: 295MB

benwtrent committed:
adding docs

benwtrent committed:
formatting

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,251 @@
+   * @param similarityFunction the similarity function used to calculate the quantile
+   * @return the new offset
+   */
+  public float recalculateCorrectiveOffset(
+      float oldOffset, ScalarQuantizer oldQuantizer, VectorSimilarityFunction similarityFunction) {
+    if (similarityFunction.equals(VectorSimilarityFunction.EUCLIDEAN)) {
+      return 0f;
+    }
+    // TODO I am not 100% sure this is correct
+    return oldOffset
+        * (minQuantile
+            / oldQuantizer.minQuantile
+            * (maxQuantile - minQuantile)
+            / (oldQuantizer.maxQuantile - oldQuantizer.minQuantile));
+  }
```
@tveasey could you double check this? I am not 100% sure of my calculation here now that we have rolled in `minQuantile * (v - minQuantile / 2.0F)`

tveasey replied:
As per discussion offline this does indeed need to be amended.

benwtrent committed:
adjusting offset correction and refactoring

benwtrent commented:
To address some of @jpountz's worries around adversarial cases, I tested one.

Cohere-Wiki, I created 100 clusters via KMeans and indexed the documents sorted by their respective cluster labels.

This ended up creating 11 total segments. Two ofthe segments needed to be requantized totally. The other 9 just needed their offsets recalculated.

HNSW Float32 Recall@10: 0.840

Quantized Recall@10: `0.787`
Recall@10|15 (did I get the true top 10 when gathering 15): `0.848`

So, we can achieve similar recall without having to requantize all vectors, even in adversarial cases. Additionally, in extreme cases, we will requantize the segment and potentially recalculate the quantiles.

benwtrent committed:
Merge branch 'main' into feature/add-scalar-quantization

benwtrent committed:
fixing up after main merge

benwtrent committed:
updating testing

benwtrent committed:
adding test coverage cleaning code

benwtrent committed:
fixing formatting

benwtrent commented:
I realize this is a large PR. I am happy to go into more details in anything that doesn't make sense (probably indicates the code isn't very readable or there should be better comments...)

I hope this is a step in the right direction, anybody ready for a review as it is now?

Requiring 4x less space with minimal recall reduction is nice :).

Pinging folks who I think would be interested (not necessarily requiring a review). 

@zhaih @jmazanec15 @msokolov @jimczi

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsFormat.java:
```diff
@@ -0,0 +1,209 @@
+ *
+ * @lucene.experimental
+ */
+public final class Lucene99HnswVectorsFormat extends KnnVectorsFormat {
```
This would be helpful for the review to get a general idea how the format is organized.

benwtrent replied:
@mayya-sharipova The format comment should contain information around how quantized vectors are stored, their sparse corrections, etc.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,782 @@
+      if (floatVectors.size() == 0) return SHALLOW_SIZE;
+      return SHALLOW_SIZE
+          + (long) floatVectors.size()
+              * (RamUsageEstimator.NUM_BYTES_OBJECT_REF + RamUsageEstimator.NUM_BYTES_ARRAY_HEADER);
```
I think we don't need `RamUsageEstimator.NUM_BYTES_ARRAY_HEADER` as here we only to account for a number of object references.

benwtrent replied:
++

jimczi commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,267 @@
+    for (int i = SCALAR_QUANTIZATION_SAMPLE_SIZE; i < numFloatVecs; i++) {
+      int j = random.nextInt(i + 1);
+      if (j < SCALAR_QUANTIZATION_SAMPLE_SIZE) {
+        vectorsToTake[j] = i;
```
nit: It's not straightforward but we could use the doc ids here to advance the iterator directly instead of iterating the vectors one by one to find the ordinal:
```
while (index <= i) {
        floatVectorValues.nextDoc();
        index++;
}
```
?
We would need to re-create the `FloatVectorValues` for the second iteration though.

benwtrent replied:
We could maybe clone this. But, advancing doc Ids via `advance(id)` isn't allowed in merged contexts (MergedVectors throws). So, iterating just the ordinals seemed easy enough.

Maybe having a separate wrapper that is passed here instead of the `MergedVectors` that doesn't do any docId translation, etc. but just iterates each vector assuming they are dense could be done. Maybe for a follow up optimization?

benwtrent replied:
Iterating the vectors directly like this is fairly cheap as we don't read from offheap unless we call `vectorValue`

jimczi replied:
Yep that's not straightforward and maybe it's not worth it. Maybe add a comment to explain why we use ordinals and cannot use advance directly.

msokolov replied:
sorry I haven't been able to review in detail, but I am disturbed that we keep around full-precision data since rather than *reducing* storage that means this will actually *increase* storage. Given that, what is the purpose of the quantization? Is it that we get faster searches? Or is this intended to be a precursor to some more radical quantization like product space quantization? I feel I'm missing the purpose.

benwtrent replied:
> . Given that, what is the purpose of the quantization? Is it that we get faster searches? Or is this intended to be a precursor to some more radical quantization like product space quantization?

using less memory on search & merge is the big benefit @msokolov but searches & merges are indeed faster by 30-40%.

Storage size increases by 25%. 

Even if some external users needed to do quantization before ingesting into Lucene, they would need to preserve the original vectors, especially if the vector data shifts.

msokolov replied:
Yes OK, this makes sense - we don't need to load the float32 vectors at search time; they can just sit on disk enabling requantization, and the storage increase is not too costly: it's not what we optimize for.

benwtrent replied:
> they can just sit on disk enabling requantization, and the storage increase is not too costly: it's not what we optimize for.

Correct. In the future @msokolov we maybe could allow users to have a "limit" where at a certain point the codec just stops storing the raw vectors. But, I think this would require a larger change as our vector search also provides doc values via `getFloatVectors`, and this isn't assumed to be lossy :/.

I would like to make that a separate and future discussion if we think users will find it useful.

jimczi reviewed (approved):
Great work @benwtrent! I like the fact that we keep the raw vectors for more flexibility in the future. We might want to expose a way to use the quantised vectors outside of the graph build/search but that's for another PR.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,782 @@
+    return BASE_RAM_BYTES_USED;
+  }
+
+  static class QuantizationVectorWriter implements Accountable {
```
Should we include `field` in this  class name to show it is responsible for a single field, for example `QuantizationFieldVectorWriter`?

benwtrent replied:
I can do that.

jmazanec15 commented:
Hey @benwtrent, sorry for delay, still looking through change. But 4x space improvement with minimal recall loss is awesome.

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,267 @@
+   * @param similarityFunction the similarity function used to calculate the quantile
+   * @return the corrective offset that needs to be applied to the score
+   */
+  public float quantize(float[] src, byte[] dest, VectorSimilarityFunction similarityFunction) {
+    assert src.length == dest.length;
+    float correctiveOffset = 0f;
+    for (int i = 0; i < src.length; i++) {
+      float v = src[i];
+      float dx = Math.max(minQuantile, Math.min(maxQuantile, src[i])) - minQuantile;
+      float dxs = scale * dx;
+      float dxq = Math.round(dxs) * alpha;
+      correctiveOffset += minQuantile * (v - minQuantile / 2.0F) + (dx - dxq) * dxq;
+      dest[i] = (byte) Math.round(dxs);
+    }
+    if (similarityFunction.equals(VectorSimilarityFunction.EUCLIDEAN)) {
+      return 0;
+    }
+    return correctiveOffset;
+  }
```
I realize this is pretty opaque. I will add some comments here about the algebra used and what we are doing with these numbers. 

Basically, we are bucketing `float` based on some upper/lower quantiles into `byte` values. In doing this, there is additionally required corrective offsets to account for linear magnitude shifts. This is why `cosine` `dot_product` and `max_inner_product` have a score corrective offset. 

For `euclidean` the algebra cancelled out all the corrective offsets and we just need a global linear one.

jmazanec15 replied:
Did this improve recall in any of the benchmarks?

benwtrent replied:
The corrective offset is required for good recall.
> Did this improve recall in any of the benchmarks?

I am unsure what you are referring to? All the benchmarks I have done have had a corrective offset.

jmazanec15 replied:
@benwtrent I was working on verifying this for my own understanding. I was able to work out the  `minQuantile * (v - minQuantile / 2.0F)` term, but Im still not sure how to get `(dx - dxq) * dxq` term.

Intuitively, I would imagine that for `f1*f2`, adding `(dx1 - dxq1) * dxq2 + (dx2 - dxq2) * dxq1` would make sense, but Im not sure where `(dx1 - dxq1) * dxq1 + (dx2 - dxq2) * dxq2` would fall out. 

For instance, if dx1=2.0 and dxq1=2.0 and dx2=2000.99 and dxq2=2000 (minQ=0 and maxQ=2000.99 for simplicity). 

```
dx1*dx2 = 4,001.98
dxq1*dxq2 = 4,000

(dx1 - dxq1) * dxq2 + (dx2 - dxq2) * dxq1 = 1.98
(dx1 - dxq1) * dxq1 + (dx2 - dxq2) * dxq2 = 1,980

dxq1*dxq2 + (dx1 - dxq1) * dxq2 + (dx2 - dxq2) * dxq1 = 4,001.98
dxq1*dxq2 + (dx1 - dxq1) * dxq1 + (dx2 - dxq2) * dxq2 = 5,980
```

Am I overlooking anything?

tveasey replied:
This comes from a linear analysis of the error terms in the dot product as a function of quantisation error `(x - lower) - (upper - lower) / n_bins * x_quantised`. This accounts for linear terms in this error in an "average" sense (where we borrow an idea from ScaNN and make the correction accurate for retrieved vectors which are similar to the query vector). We plan to write up the idea behind this approach in full elsewhere. In fact, for 7 bit quantisation the effect on quality of retrieval is relatively small, but for fewer bits we found it significantly improved recall.

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
adding many comments

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,782 @@
+    float upperQuantile = 0f;
+    int totalCount = 0;
+    for (int i = 0; i < quantizationStates.size(); i++) {
+      if (quantizationStates.get(i) == null) {
```
Should this attempt to merge the others or is it not worth it?

benwtrent replied:
Since we have no knowledge of the vectors or their quantiles from this segment, its better to require us to recalculate the quantiles and requantize.

We could continue merging the rest, but in the end, we have no information around how different the other vectors are. I think its good to play it safe here.

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,782 @@
+  private static final long BASE_RAM_BYTES_USED =
+      shallowSizeOfInstance(Lucene99ScalarQuantizedVectorsWriter.class);
+
+  private static final float QUANTIZATION_RECOMPUTE_LIMIT = 32;
```
What was reasoning behind choosing this number?

benwtrent replied:
@jmazanec15 I added comments for both this number and the `0.2f` to describe the reasoning. Both were determined via empirical testing with adverse and typical use cases (cohere and e5 embeddings, sorted via magnitude and sorted via clusters).

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,782 @@
+
+  static boolean shouldRequantize(ScalarQuantizer existingQuantiles, ScalarQuantizer newQuantiles) {
+    // Should this instead be 128f?
+    float tol = 0.2f * (newQuantiles.getUpperQuantile() - newQuantiles.getLowerQuantile()) / 128f;
```
What was reasoning for 0.2?

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,316 @@
+ * Will scalar quantize float vectors into `int8` byte values. This is a lossy transformation.
+ * Scalar quantization works by first calculating the quantiles of the float vector values. The
+ * quantiles are calculated using the configured quantile/confidence interval. The [minQuantile,
+ * maxQuantile] are then used to scale the values into the range [0, 127] and bucketed into the
```
Maybe simple question: why not [0, 255]?

benwtrent replied:
@jmazanec15 Java doesn't have unsigned bytes & handling signed values in the math is exceptionally difficult. 

We only use one bucket of possible values, the impact in recall is minimal (if measurable at all).

benwtrent replied:
@jmazanec15 I think we could increase it to `[0, 255]`, but it would require new similarity methods in VectorUtil, specifically `unsignedDotProduct` and `unsignedCosine`, but that shouldn't be too bad. 

I can do it as a follow up to this PR.

benwtrent replied:
@jmazanec15 https://github.com/apache/lucene/pull/12694

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,316 @@
+ * maxQuantile].
+ *
+ * <pre class="prettyprint">
+ *   float = (maxQuantile - minQuantile)/127 * byte + minQuantile
```
nit: for readability, might be good to switch L42 and L43. Quantize the float, then reconstruct it.

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,316 @@
+ * following:
+ *
+ * <pre class="prettyprint">
+ *   float1 * float2 = (byte1 * (maxQuantile - minQuantile)/127 + minQuantile) * (byte2 * (maxQuantile - minQuantile)/127 + minQuantile)
```
nit: could we change this to "~=" instead of "=" to emphasize loss.

benwtrent replied:
++

uschindler commented:
Hi, why do we need a new Codec? The Lucebe main file format does not change, olly the HNSW format was exchanged. Because like pistingsfornats and dicvaluesformats, the SPI can detect the format of the HNSW index by reading the file.
The top level Codec itself does not need to be versionized. You just need to change its defaults when it writes new segments.
Basically the Lucene95 codec was also possibly a useless addition because the top level Codec did not change at all

jmazanec15 commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,316 @@
+ * maxQuantile] are then used to scale the values into the range [0, 127] and bucketed into the
+ * nearest byte values.
+ *
+ * <h2>How Scalar Quantization Works</h2>
```
question: Where does the configured quantile come into play for these calculations?

benwtrent replied:
@jmazanec15 the configured quantile/confidence interval is used to calculate the upper and lower quantiles for the accepted values. I think earlier I mention this in the comment.

jimczi commented:
>  why do we need a new top-level Codec? The Lucene main file format does not change, only the HNSW format was exchanged. Because like ppostingsfornats and docvaluesformats, the SPI can detect the format of the HNSW index by reading the file and uses SPI to lookup the correct format.

That's a good point. I think we'd need to increment the VERSION_CURRENT of the Lucene95HnswVectorsFormat to do the right thing when reading the data and we could avoid the new format entirely since it's exactly the same as before (assuming that quantisation is disabled by default).

uschindler commented:
> > why do we need a new top-level Codec? The Lucene main file format does not change, only the HNSW format was exchanged. Because like ppostingsfornats and docvaluesformats, the SPI can detect the format of the HNSW index by reading the file and uses SPI to lookup the correct format.
> 
> That's a good point. I think we'd need to increment the VERSION_CURRENT of the Lucene95HnswVectorsFormat to do the right thing when reading the data and we could avoid the new format entirely since it's exactly the same as before (assuming that quantisation is disabled by default).

Actually, if the HNSW format has its own SPI name, when reading indexes it should be chosen automatically by KNNVectorsFormat.forName(): https://lucene.apache.org/core/9_0_0/core/org/apache/lucene/codecs/KnnVectorsFormat.html?is-external=true#forName(java.lang.String)

In short: when top level codec reads the index and opens the vector format it would read the SPI name header from file and then load the correct code (possibly the actual one or knew from backwards). That's working like that for years for postings and docvalues.

uschindler commented:
> @uschindler so I should just add a new format?
> 
> It would be a new Lucene99 HNSW format, but keep the default Lucene95 HNSW format?
> 
> Or can we change the default vector format for the top level Lucene95 codec to a new HNSW format?
> 
> I definitely don't fully understand how backwards compatibility works.

No you just create a new HNSW format with a new name. The old one is moved to backwards compatibility.

The top level Codec Lucene95Codec ist Format compatible and needs no change in it's identify. Only the part that returns the format for writing must be changed.

When it reads an index the format is looked up by its name from the metadata. If that does not work there is a bug in the code reading the segment Infos.

So basically, only add a new HNSW format (like you could add a new postings format) but leave the codec intact. Just change the class returned for HNSW when writing.

The top level codec only needs to be changed when metadata fotmats or non spi formats change.

uschindler commented:
I just checked the code, the 9.5 top-level codec addition was useless. Just code duplication. We can't revert it anymore, but we should not repeat that. The only required top-level Format for 9.x is Lucene90Codec. It was duplicated for no reason.

uschindler commented:
The simplest change is:
- Remove Lucene99Codec
- In Lucene95Codec just change this: `this.defaultKnnVectorsFormat = new Lucene95HnswVectorsFormat();` to the new format.

Done. 

(when index is written it will write the name of the codec into FieldInfos; when reading the Lucene 95 codec will read the name and looks it up to load the format-sepecific format class)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsWriter.java:
```diff
@@ -0,0 +1,1149 @@
+        assert scalarQuantizer != null;
+        quantizedVectorDataOffsetAndLength = new long[2];
+        quantizedVectorDataOffsetAndLength[0] = quantizedVectorData.alignFilePointer(Float.BYTES);
+        scorerSupplier =
```
Does this mean that we will use quantized vectors for graph building during merging?

benwtrent replied:
> Does this mean that we will use quantized vectors for graph building during merging?

Yes we will.

> which brings up another question: where are raw vector values are used?

If the quantiles need to be rebuilt. Additionally raw vectors are used when reading floats directly from the KnnVectorReader.

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
addressing PR comment

benwtrent committed:
fixing test and clarifying quantile check values

benwtrent commented on lucene/core/src/java/org/apache/lucene/util/ScalarQuantizer.java:
```diff
@@ -0,0 +1,317 @@
+ */
+public class ScalarQuantizer {
+
+  public static final int SCALAR_QUANTIZATION_SAMPLE_SIZE = 25_000;
```
This is empirical and fairly conservative (on the side of getting good quantiles). It probably could be configurable, but I went for simple first with sane defaults.

benwtrent commented:
@uschindler I switch it around. Lucene95Codec is back, moved the previous HNSW format into backwards_codec and switched Lucene95Codec to use the Lucene99Hnsw format.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsFormat.java:
```diff
@@ -0,0 +1,231 @@
+    super("Lucene99HnswVectorsFormat");
+    if (maxConn <= 0 || maxConn > MAXIMUM_MAX_CONN) {
+      throw new IllegalArgumentException(
+          "maxConn must be postive and less than or equal to"
```
nit: "postive" => "positive"

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsFormat.java:
```diff
@@ -0,0 +1,231 @@
+    }
+    if (beamWidth <= 0 || beamWidth > MAXIMUM_BEAM_WIDTH) {
+      throw new IllegalArgumentException(
+          "beamWidth must be postive and less than or equal to"
```
nit: "postive" => "positive"

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsReader.java:
```diff
@@ -0,0 +1,628 @@
+    }
+    if (quantizedVectorsReader != null) {
+      OffHeapQuantizedByteVectorValues vectorValues = getQuantizedVectorValues(field);
+      if (vectorValues == null) {
```
if there are not quantized vector values, should we do search with raw vector values?  (for vectors fields where quantization is not enabled)

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsReader.java:
```diff
@@ -0,0 +1,628 @@
+        || fieldEntry.vectorEncoding != VectorEncoding.FLOAT32) {
+      return;
+    }
+    if (quantizedVectorsReader != null) {
```
Should we instead check if `if (fieldEntry.hasQuantizedVectors())`?

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99HnswVectorsReader.java:
```diff
@@ -0,0 +1,628 @@
+  @Override
+  public OffHeapQuantizedByteVectorValues getQuantizedVectorValues(String field)
+      throws IOException {
+    FieldEntry fieldEntry = fields.get(field);
```
would be nice not to search `fields` Map and do these extra checks when getting quantized vectors for search.

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsWriter.java:
```diff
@@ -0,0 +1,824 @@
+          scalarQuantizer.quantize(v, vector, fieldData.vectorSimilarityFunction);
+      quantizedVectorData.writeBytes(vector, vector.length);
+      offsetBuffer.putFloat(offsetCorrection);
+      quantizedVectorData.writeBytes(offsetBuffer.array(), offsetBuffer.array().length);
```
nit: We can just use `quantizedVectorData.writeInt(Float.floatToIntBits(offsetCorrection))` without using `offsetBuffer`

mayya-sharipova commented on lucene/core/src/java/org/apache/lucene/codecs/lucene99/OffHeapQuantizedByteVectorValues.java:
```diff
@@ -0,0 +1,275 @@
+    }
+    slice.seek((long) targetOrd * byteSize);
+    slice.readBytes(byteBuffer.array(), byteBuffer.arrayOffset(), dimension);
+    slice.readFloats(scoreCorrectionConstant, 0, 1);
```
nit: here, instead we can do:
` float scoreCorrectionConstant = Float.intBitsToFloat(slice.readInt());`

mayya-sharipova reviewed (approved):
@benwtrent Amazing work!!! Thanks Ben, Jim and Tom for such a great contribution! 
I left some small comments, but otherwise this PR LGTM (I am still fuzzy about all the math in the quantizer, so I trust you and @tveasey that you got it right).

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent committed:
addressing PR comments

benwtrent committed:
Merge remote-tracking branch 'upstream/main' into feature/add-scalar-quantization

benwtrent commented:
@naveentatikonda , I used lucene-util.

It's an off-heap memory requirement reduction. Instead of having to load float32 values in memory, it will load byte values. HNSW does many random reads, this read behavior really needs to be in memory for it to be fast.

It uses about 25% more disk as the raw float32 vectors are kept along with the quantized ones.

navneet1v commented:
> @naveentatikonda , I used lucene-util.
> 
> It's an off-heap memory requirement reduction. Instead of having to load float32 values in memory, it will load byte values. HNSW does many random reads, this read behavior really needs to be in memory for it to be fast.
> 
> It uses about 25% more disk as the raw float32 vectors are kept along with the quantized ones.

can you share some code reference and dataset you used for testing? It will be very helpful.

benwtrent commented:
https://github.com/mikemccand/luceneutil

It can be a little frustrating at first to set up, but it's nice for quick integration and performance tests.

The python script knnPerfTest.py and I downloaded a bunch of CohereV2 parquet files from huggingface. 

CohereV3 is released now and they have many GB of vectors available for testing on huggingface. So, you could use those instead.

Love those folks over at Cohere :)

navneet1v commented:
Thanks @benwtrent will look into this.
